pkg/controller/persistentvolumelabel/pvlcontroller.go

```diff
@@ -0,0 +1,289 @@
+	// available, so we don't worry about secondary waits
+	for pvlc.processNextWorkItem() {
+	}
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (pvlc *PersistentVolumeLabelController) processNextWorkItem() bool {
+	// pull the next work item from queue.  It should be a key we use to lookup something in a cache
+	key, quit := pvlc.queue.Get()
+	glog.Info("Pulled PV work from queue")
+	if quit {
+		return false
+	}
+	// you always have to indicate to the queue that you've completed a piece of work
+	defer pvlc.queue.Done(key)
+
+	// do your work on the key.  This method will contains your "do stuff" logic
+	err := pvlc.syncHandler(key.(*v1.PersistentVolume))
+	if err == nil {
+		// if you had no error, tell the queue to stop tracking history for your key.  This will
+		// reset things like failure counts for per-item rate limiting
+		pvlc.queue.Forget(key)
+		return true
+	}
+
+	// there was a failure so be sure to report it.  This method allows for pluggable error handling
+	// which can be used for things like cluster-monitoring
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
+
+	// since we failed, we should requeue the item to work on later.  This method will add a backoff
+	// to avoid hotlooping on particular items (they're probably still not going to work right away)
+	// and overall controller protection (everything I've done is broken, this controller needs to
+	// calm down or it can starve other useful work) cases.
+	pvlc.queue.AddRateLimited(key)
+
+	return true
+}
+
+// AddLabels adds appropriate labels to persistent volumes and sets the
+// volume as available if successful.
+func (pvlc *PersistentVolumeLabelController) AddLabels(volume *v1.PersistentVolume) error {
+	glog.Info("Adding PV Labels")
+	var volumeLabels map[string]string
+	if volume.Spec.AWSElasticBlockStore != nil {
+		labels, err := pvlc.findAWSEBSLabels(volume)
+		if err != nil {
+			return fmt.Errorf("error querying AWS EBS volume %s: %v", volume.Spec.AWSElasticBlockStore.VolumeID, err)
+		} else {
+			volumeLabels = labels
+		}
+	}
+	if volume.Spec.GCEPersistentDisk != nil {
+		labels, err := pvlc.findGCEPDLabels(volume)
+		if err != nil {
+			return fmt.Errorf("error querying GCE PD volume %s: %v", volume.Spec.GCEPersistentDisk.PDName, err)
+		} else {
+			volumeLabels = labels
+		}
+	}
+
+	glog.Infof("Found %d labels: %v", len(volumeLabels), volumeLabels)
+	if len(volumeLabels) != 0 {
+		if volume.Labels == nil {
+			volume.Labels = make(map[string]string)
+		}
+		for k, v := range volumeLabels {
+			// We (silently) replace labels if they are provided.
+			// This should be OK because they are in the kubernetes.io namespace
+			// i.e. we own them
+			glog.Infof("Adding label %s=%s", k, v)
+			volume.Labels[k] = v
+		}
+		//	glog.Info("Pausing before setting volume available")
+		//	time.Sleep(30 * time.Second)
+		volume.Status.Phase = v1.VolumeAvailable
```

Comment:
(I feel like I've done this review before, so forgive me if you've already explained this somewhere and just point me there!)

So it is possible that there are no labels, and I don't understand why we wouldn't then mark the volume as Available in that case also.  But in general, what is the lifecycle of the volume phase here?  It looks like we aren't solely responsible for updating the phase.  And in that case,  are we guaranteed that the volume won't be used before we label it?